qa/rpc-tests/overwinter_peer_management.py

```diff
@@ -0,0 +1,115 @@
+#!/usr/bin/env python2
+# Copyright (c) 2018 The Zcash developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+from test_framework.mininode import NodeConn, NodeConnCB, NetworkThread, \
+    EarlyDisconnectError, msg_inv, mininode_lock, msg_ping, \
+    MY_VERSION, OVERWINTER_PROTO_VERSION
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import initialize_chain_clean, start_nodes, \
+    p2p_port, assert_equal
+
+import time
+
+#
+# In this test we connect Sprout and Overwinter mininodes to a Zcashd node
+# which will activate Overwinter at block 10.
+#
+# We test:
+# 1. the mininodes stay connected to Zcash with Sprout consensus rules
+# 2. when Overwinter activates, the Sprout mininodes are dropped
+# 3. new Overwinter nodes can connect to Zcash
+# 4. new Sprout nodes cannot connect to Zcash
+#
+# This test *does not* verify that prior to Overwinter activation, the Zcashd
+# node will prefer connections with Overwinter nodes, with an eviction process
+# that prioritizes Sprout connections.
+#
+
+
+class TestManager(NodeConnCB):
+    def __init__(self):
+        NodeConnCB.__init__(self)
+        self.create_callback_map()
+
+    def on_close(self, conn):
+        pass
+
+    def on_reject(self, conn, message):
+        conn.rejectMessage = message
+
+
+class OverwinterPeerManagementTest(BitcoinTestFramework):
+
+    def setup_chain(self):
+        print "Initializing test directory "+self.options.tmpdir
+        initialize_chain_clean(self.options.tmpdir, 1)
+
+    def setup_network(self):
+        self.nodes = start_nodes(1, self.options.tmpdir,
+                                 extra_args=[['-nuparams=5ba81b19:10', '-debug', '-whitelist=127.0.0.1']])
+
+    def run_test(self):
+        test = TestManager()
+
+        # Launch 10 Sprout and 10 Overwinter Nodes
+        nodes = []
+        for x in xrange(10):
+            nodes.append(NodeConn('127.0.0.1', p2p_port(0), self.nodes[0], test, "regtest", False))
+            nodes.append(NodeConn('127.0.0.1', p2p_port(0), self.nodes[0], test, "regtest", True))
+
+        # Start up network handling in another thread
+        NetworkThread().start()
+
+        # Sprout consensus rules apply after mining 9 blocks
+        self.nodes[0].generate(9)
```

Comment:
I think there's an off-by-one error. Per ZIP 200, `ACTIVATION_HEIGHT - 1` is the last common block, so once we receive that block, we don't want to receive the next block that non-upgrading nodes broadcast.

Reply:
I don't think there's an off-by-one.  Just the test comments were not clear.  The zcash node in `self.nodes[0`] has an activation height of 10.  At block height 9, the test confirms that Sprout and Overwinter mininodes (which advertise version 170002 and 170003) are all connected to the zcash node. The mininode connections don't care about blocks; only whether or not the mininode can connect and/or remain connected.